File: data_structures/week1_basic_data_structures/stack_with_max/stack_with_max_naive.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <string>
#include <vector>

class StackWithMax {
  std::vector<int> stack;

 public:
  void Push(int value) { stack.push_back(value); }

  void Pop() {
    assert(stack.size());
    stack.pop_back();
  }

  int Max() const {
    assert(stack.size());
    return *std::max_element(stack.begin(), stack.end());
  }
};

int main() {
  int num_queries = 0;
  std::cin >> num_queries;

  std::string query;
  std::string value;

  StackWithMax stack;

  for (int i = 0; i < num_queries; ++i) {
    std::cin >> query;
    if (query == "push") {
      std::cin >> value;
      stack.Push(std::stoi(value));
    } else if (query == "pop")
      stack.Pop();
    else if (query == "max")
      std::cout << stack.Max() << std::endl;
    else
      assert(0);
  }
  return 0;
}

File: data_structures/week1_basic_data_structures/stack_with_max/stack_with_max_naive.py
from __future__ import annotations

import sys
from typing import Generic, List, TypeVar

T = TypeVar("T", int, float)


class StackWithMax(Generic[T]):
    def __init__(self) -> None:
        self.__stack: List[T] = []

    def push(self, a: T) -> None:
        self.__stack.append(a)

    def pop(self) -> None:
        assert len(self.__stack)

        self.__stack.pop()

    def max(self) -> T:  # noqa: A003
        assert len(self.__stack)

        return max(self.__stack)


if __name__ == "__main__":  # noqa: C901
    stack: StackWithMax[int] = StackWithMax()

    n = int(sys.stdin.readline())
    for _ in range(n):
        query = sys.stdin.readline().split()

        if query[0] == "push":
            stack.push(int(query[1]))
        elif query[0] == "pop":
            stack.pop()
        elif query[0] == "max":
            print(stack.max())
